### PR TITLE
fix #31521, make regexes thread-safe

### LINKED ISSUE #31521: regex not threadsafe
There's a race condition for creating the regex JIT cache from different threads.

### PR DESCRIPTION
**Pull request by JeffBezanson**

### CODE REVIEW

File: base/regex.jl [hunk 1]
```diff
@@ -384,9 +387,23 @@ julia> replace(msg, r"#(.+)# from (?<from>\\w+)" => s"FROM: \\g<from>; MESSAGE:
 """
 macro s_str(string) SubstitutionString(string) end
 
+# replacement
+
+struct RegexAndMatchData
+    re::Regex
+    match_data::Ptr{Cvoid}
+    RegexAndMatchData(re::Regex) = (compile(re); new(re, PCRE.create_match_data(re.regex)))
```
Comment (vtjnash) [inline base/regex.jl]:
```suggestion
    function RegexAndMatchData(re::Regex)
         compile(re)
         r = new(re, PCRE.create_match_data(re.regex))
         finalizer(_free_pat_replacer, r)
         return r
    end
```

File: base/regex.jl [hunk 2]
```diff
@@ -384,9 +387,23 @@ julia> replace(msg, r"#(.+)# from (?<from>\\w+)" => s"FROM: \\g<from>; MESSAGE:
 """
 macro s_str(string) SubstitutionString(string) end
 
+# replacement
+
+struct RegexAndMatchData
```
Comment (vtjnash) [inline base/regex.jl]:
```suggestion
mutable struct RegexAndMatchData
```

File: base/regex.jl [hunk 3]
```diff
@@ -384,9 +387,23 @@ julia> replace(msg, r"#(.+)# from (?<from>\\w+)" => s"FROM: \\g<from>; MESSAGE:
...
+    re::Regex
+    match_data::Ptr{Cvoid}
+    RegexAndMatchData(re::Regex) = (compile(re); new(re, PCRE.create_match_data(re.regex)))
+end
+
+findnext(pat::RegexAndMatchData, str, i) = _findnext_re(pat.re, str, i, pat.match_data)
+
+_pat_replacer(r::Regex) = RegexAndMatchData(r)
+
+_free_pat_replacer(r::RegexAndMatchData) = PCRE.free_match_data(r.match_data)
```
Comment (vtjnash) [inline base/regex.jl]:
```suggestion
function _free_pat_replacer(r::RegexAndMatchData)
    data = r.match_data
    r.match_data = C_NULL
    data == C_NULL || PCRE.free_match_data(data)
    nothing
end
```

**Review by vtjnash — APPROVED** [review]
I'm suggesting adding a finalizer to RegexAndMatchData so that there isn't a memory leak if there's an error throw.

How's the performance of calling malloc and free here? @nanosoldier `runbenchmarks("parse_json" || "regex_dna" || "regex_dna", vs=":master")`

**Comment by JeffBezanson** [comment]
@nanosoldier `runbenchmarks("parse_json" || "regex_dna" || "regex_dna", vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/32381#issuecomment-504524722) has completed - no performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/ca61164c14c3e5dfb7317d521d1a6f8b563a2d86/136d75c_vs_75c10e4/report.md). cc @ararslan

**Pull request merged**